sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1122 @@
+# coding: utf-8
+"""
+Large Margin Nearest Neighbor Classification
+"""
+
+# Author: John Chiotellis <johnyc.code@gmail.com>
+# License: BSD 3 clause
+
+from __future__ import print_function
+from warnings import warn
+
+import sys
+import time
+import numpy as np
+from scipy.optimize import minimize
+from scipy.sparse import csr_matrix, csc_matrix, coo_matrix, spdiags
+
+from ..base import BaseEstimator, TransformerMixin
+from ..neighbors import NearestNeighbors
+from ..decomposition import PCA
+from ..utils import gen_batches
+from ..utils.extmath import row_norms, safe_sparse_dot
+from ..utils.random import check_random_state
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import check_is_fitted, check_array, check_X_y
+
+
+class LargeMarginNearestNeighbor(BaseEstimator, TransformerMixin):
+    """Distance Metric Learning for Large Margin Classification.
+
+    Parameters
+    ----------
+    n_neighbors : int, optional (default=3)
+        Number of neighbors to use as target neighbors for each sample.
+
+    targets_algorithm : str {'auto', 'ball_tree', 'kd_tree', 'brute'}, optional
+        Algorithm used to compute the target neighbors, passed to a
+        :class:`neighbors.NearestNeighbors` instance.
+
+    init : string or numpy array, optional (default='pca')
+        Initialization of the linear transformation.
+        Possible options are 'pca', 'identity' and a numpy array of shape
+        (n_features_out, n_features).
```

Comment:
Perhaps note that `identity` may be truncated if `n_features_out` is specified.

Reply:
Sure.